parsec/interfaces/ptg/ptg-compiler/jdf2c.c

```diff
@@ -6741,54 +6750,69 @@ static void jdf_generate_code_hook_gpu(const jdf_t *jdf,
     coutput("  return PARSEC_HOOK_RETURN_DONE;\n"
             "}\n\n");
 
+    info.sa = string_arena_new(64);
+    info.prefix = "";
+    info.suffix = "";
+    info.assignments = "&this_task->locals";
+    /* Generate the evaluate if a device type is set */
+    /* Get the hint for static and/or external gpu scheduling */
+    jdf_find_property( body->properties, "device", &device_property );
+    if ( NULL != device_property ) {
+        device_expr = dump_expr((void**)device_property->expr, &info);
+        coutput("static int eval_%s_%s_select_device(%s *this_task)\n"
+                "{\n"
+                "  __parsec_%s_internal_taskpool_t *__parsec_tp = (__parsec_%s_internal_taskpool_t *)this_task->taskpool; (void)__parsec_tp;\n"
+                "  int dev_prop = %s;\n"
+                "  int nb_devs_of_type, dev_index;\n"
+                "  if( NULL != this_task->selected_device ) {\n"
+                "    char tmp[MAX_TASK_STRLEN];\n"
+                "    parsec_warning(\"Task %%s property device (value %%d) has been ignored \"\n"
+                "                   \"because the task already has a selected device %%d %%s\",\n"
+                "       parsec_task_snprintf(tmp, MAX_TASK_STRLEN, (parsec_task_t *)this_task), dev_prop,\n"
+                "       this_task->selected_device->device_index, this_task->selected_device->name);\n"
+                "    return PARSEC_HOOK_RETURN_DONE;\n"
+                "  }\n"
+                "  for( dev_index = nb_devs_of_type = 0; dev_index < parsec_mca_device_enabled(); dev_index++ )\n"
```

Comment:
This code should have been into a parsec internal function. That would have simplified the entire code, because you could have just dumped the device property as a function, then call use it's pointer to call it in validation function, before doing the checks below.